llvm/lib/Target/AIE/AIEPostPipeliner.h

```diff
@@ -43,14 +44,65 @@ class NodeInfo {
   int ModuloCycle = 0;
   // Cycle / II
   int Stage = 0;
+
+  int StaticEarliest = 0;
+  int StaticLatest = -1;
+
   // The earliest cycle at which this can be scheduled to meet latencies
   // This includes the lowerbound of the modulo condition, i.e.
   // Earliest(N) >= Cycle(N - NInstr) + II
   int Earliest = 0;
-  // For an LCD K1 -> K2, this holds II + Earliest(K2 - NInstr) - Latency(LCD)
-  // Instructions with lower Latest have higher priority in the
-  // top down scheduling
-  int Latest = 0;
+
+  // The latest cycle at which this can be scheduled. This is a negative value
+  // relative to the length of the linear schedule.
+  int Latest = -1;
+
+  // Record critical path components
+  // The Pred/Succ that pushed my Earliest/Latest
+  std::optional<int> LastEarliestPusher;
+  std::optional<int> LastLatestPusher;
+  // The number of Succs/Preds whose Earliest/Latest I have pushed.
+  int NumPushedEarliest = 0;
+  int NumPushedLatest = 0;
+
+  // Latest corrected taking Earliest of an LCD successor into account
+  int LCDLatest = -1;
+
+  // The transitive closure of my predecessors
+  std::unordered_set<int> Ancestors;
+
+  // The transitive closure of my successors
+  std::unordered_set<int> Offspring;
+
+  /// Reset the node to the values computed statically
+  /// If FullReset is true, also reset the accumulated dynamic data.
+  void reset(bool FullReset);
+};
+
+class PostPipelinerStrategy {
+protected:
+  ScheduleDAGInstrs &DAG;
+  std::vector<NodeInfo> &Info;
+  int LatestBias = 0;
+
+public:
+  PostPipelinerStrategy(ScheduleDAGInstrs &DAG, std::vector<NodeInfo> &Info,
+                        int LatestBias)
+      : DAG(DAG), Info(Info), LatestBias(LatestBias) {};
+  virtual ~PostPipelinerStrategy() {};
+  virtual std::string name() { return "PostPipelinerStrategy"; }
+  // Choose among available alternatives
+  virtual bool better(const SUnit &A, const SUnit &B) { return false; }
+  // Tweak the effective earliest
+  virtual int earliest(const SUnit &N) { return Info[N.NodeNum].Earliest; }
```

Comment:
Nit: The docstring makes it seem like `earliest()` is tweaking the value. Instead, I believe it is return a value that has already been tweaked. Could this be `const`?

Same for `latest()` below.

Reply:
Well, these default implementation could be const, but I don't want to block a scheme where some user state  is maintained.